pkg/querier/querier.go

```diff
@@ -154,27 +154,97 @@ func (q *V3ioQuerier) queryNumericPartition(
 }
 
 // return the current metric names
-func (q *V3ioQuerier) LabelValues(name string) ([]string, error) {
+func (q *V3ioQuerier) LabelValues(labelKey string) ([]string, error) {
+	if labelKey == "__name__" {
+		return q.getMetricNames()
+	} else {
+		return q.getLabelValues(labelKey)
+	}
+}
+
+func (q *V3ioQuerier) Close() error {
+	return nil
+}
+
+func (q *V3ioQuerier) getMetricNames() ([]string, error) {
+	input := v3io.GetItemsInput{
+		Path:           q.cfg.Path + "/names/",
+		AttributeNames: []string{"__name"},
+	}
 
-	list := []string{}
-	input := v3io.GetItemsInput{Path: q.cfg.Path + "/names/", AttributeNames: []string{"__name"}, Filter: ""}
 	iter, err := utils.NewAsyncItemsCursor(q.container, &input, q.cfg.QryWorkers, []string{}, q.logger)
-	q.logger.DebugWith("GetItems to read names", "input", input, "err", err)
 	if err != nil {
-		return list, err
+		return nil, err
 	}
 
+	var metricNames []string
+
 	for iter.Next() {
-		name := iter.GetField("__name").(string)
-		list = append(list, name)
+		metricNames = append(metricNames, iter.GetField("__name").(string))
 	}
 
 	if iter.Err() != nil {
-		q.logger.InfoWith("Failed to read names, assume empty list", "err", iter.Err().Error())
+		q.logger.InfoWith("Failed to read metric names, returning empty list", "err", iter.Err().Error())
 	}
-	return list, nil
+
+	return metricNames, nil
 }
 
-func (q *V3ioQuerier) Close() error {
-	return nil
+func (q *V3ioQuerier) getLabelValues(labelKey string) ([]string, error) {
+
+	// sync partition manager (hack)
+	err := q.partitionMngr.ReadAndUpdateSchema()
+	if err != nil {
+		return nil, err
+	}
+
+	// if no partitions yet - there are no labels
+	if len(q.partitionMngr.GetPartitionsPaths()) == 0 {
```

Comment:
You better keep the result in a local variable. Despite its name, this method is not a simple getter.

Reply:
I don't understand why a local variable is needed. Behind the scenes the return value is kept on the stack, equivalent to a local variable